driver-core/src/main/com/mongodb/connection/InternalStreamConnection.java

```diff
@@ -259,12 +263,37 @@ public void sendMessageAsync(final List<ByteBuf> byteBuffers, final int lastRequ
 
     @Override
     public void receiveMessageAsync(final int responseTo, final SingleResultCallback<ResponseBuffers> callback) {
-        notNull("open", stream, callback);
+        isTrue("stream is open", stream != null, callback);
+
         if (LOGGER.isTraceEnabled()) {
             LOGGER.trace(format("Queuing read message: %s. ([%s] %s)", responseTo, getId(), serverId));
         }
-        readQueue.put(responseTo, errorHandlingCallback(callback, LOGGER));
-        processPendingReads();
+
+        if (isClosed()) {
+            callback.onResult(null, new MongoSocketClosedException("Can not read from a closed socket", getServerAddress()));
+        }
+
+        readerLock.lock();
+
+        Response response = messages.get(responseTo);
+
+        try {
+            if (response == null) {
+                readQueue.put(responseTo, callback);
+            }
+
+            if (!readQueue.isEmpty() && !isReading) {
```

Comment:
This conditional is the one I'm least sure about.  The second condition, I believe, is necessary to prevent simultaneous readers.    And without the first condition the "InternalStreamConnectionSpecification#should handle out of order messages on the stream asynchronously" test fails on the second async read.